cpp/include/cudf/io/csv.hpp

```diff
@@ -0,0 +1,1447 @@
+  char get_thousands() const { return _thousands; }
+
+  /**
+   * @brief Returns decimal point character.
+   */
+  char get_decimal() const { return _decimal; }
+
+  /**
+   * @brief Returns comment line start character.
+   */
+  char get_comment() const { return _comment; }
+
+  /**
+   * @brief Whether to treat `\r\n` as line terminator.
+   */
+  bool is_enabled_windowslinetermination() const { return _windowslinetermination; }
+
+  /**
+   * @brief Whether to treat whitespace as field delimiter.
+   */
+  bool is_enabled_delim_whitespace() const { return _delim_whitespace; }
+
+  /**
+   * @brief Whether to skip whitespace after the delimiter.
+   */
+  bool is_enabled_skipinitialspace() const { return _skipinitialspace; }
+
+  /**
+   * @brief Whether to ignore empty lines or parse line values as invalid.
+   */
+  bool is_enabled_skip_blank_lines() const { return _skip_blank_lines; }
+
+  /**
+   * @brief Returns quoting style.
+   */
+  quote_style get_quoting() const { return _quoting; }
+
+  /**
+   * @brief Returns quoting character.
+   */
+  char get_quotechar() const { return _quotechar; }
+
+  /**
+   * @brief Whether a quote inside a value is double-quoted.
+   */
+  bool is_enabled_doublequote() const { return _doublequote; }
+
+  /**
+   * @brief Returns names of columns to read as datetime.
+   */
+  std::vector<std::string> const& get_infer_date_names() const { return _infer_date_names; }
+
+  /**
+   * @brief Returns indexes of columns to read as datetime.
+   */
+  std::vector<int> const& get_infer_date_indexes() const { return _infer_date_indexes; }
+
+  /**
+   * @brief Returns per-column types.
+   */
+  std::vector<std::string> const& get_dtypes() const { return _dtypes; }
+
+  /**
+   * @brief Returns additional values to recognize as boolean true values.
+   */
+  std::vector<std::string> const& get_true_values() const { return _true_values; }
+
+  /**
+   * @brief Returns additional values to recognize as boolean false values.
+   */
+  std::vector<std::string> const& get_false_values() const { return _false_values; }
+
+  /**
+   * @brief Returns additional values to recognize as null values.
+   */
+  std::vector<std::string> const& get_na_values() const { return _na_values; }
+
+  /**
+   * @brief Whether to keep the built-in default NA values.
+   */
+  bool is_enabled_keep_default_na() const { return _keep_default_na; }
+
+  /**
+   * @brief Whether to disable null filter.
+   */
+  bool is_enabled_na_filter() const { return _na_filter; }
+
+  /**
+   * @brief Whether to parse dates as DD/MM versus MM/DD.
+   */
+  bool is_enabled_dayfirst() const { return _dayfirst; }
+
+  /**
+   * @brief Returns timestamp_type to which all timestamp columns will be cast.
+   */
+  data_type get_timestamp_type() const { return _timestamp_type; }
+
+  /**
+   * @brief Sets compression format of the source.
+   *
+   * @param comp Compression type.
+   */
+  void set_compression(compression_type comp) { _compression = comp; }
```

Comment:
Can we make these settings/mutable functions private and access them via friend?

Reply:
I've been thinking about this one too, and still prefer the public API here. IMO the options class should be usable without the builder.
I'm not 100% sure if this argument is valid in the builder pattern case but haven't found any examples online to point in either direction.

Reply:
Builders provide us with a couple advantages, one being that we can have a context fluent api that allows us to set options in an intuitive way while still allowing the consumer to grok the naming conventions. Another being that the options can't change once created.
```c++
auto car_builder = car_builder()
  .wheels(4)
  .doors(2)
  .sunroof(true)
  .drive_modes("4x4");
```
builder keeps ownership until build, when options are either moved or copied out of the builder.
```c++
auto car_config = car_builder.build();
auto num_wheels = car_config.get_num_wheels(); // 2
auto has_sunroof = car_config.has_sunroof(); // true
```
the options themselves should not be mutable
```c++ 
car_config.set_num_wheels(7); // should compile
```
but can be "updated" immutably by creating another builder which copies or moves the original
```c++
auto con_config_updated = car_builder(std::move(car_config))
  .wheels(18)
  .sunroof(false)
  .build();

auto num_wheels_2 = car_config_updated.get_num_wheels(); // 18
auto has_sunroom_2 = car_config_updated.get_num_wheels(); // true
```
This is important in other languages (when encapsulation is the primary strategy for maintaining immutability), but less so in C++ because we can nearly enforce immutability by making the options `const`.
```c++
auto car_config_1 = car_builder().wheels(4).build();
car_config_1.set_num_wheels(7); // ok

auto const car_config_2 = car_builder().wheels(4).build();
car_config_2.set_num_wheels(7); // wont compile, because const.
```
So mutability wise, it's not critical to avoid the mutable functions on the options. But that begs the question, why have the builder at all?

Reply:
Yes, I chain-ability is one of the feature that we are taking advantage of in builder API. And having public set api from main class will provide flexibility of updating the object later down the line (but the chances of updating is quite less).

Reply:
Updating the options after they’ve been built is a code smell to me. I’d there an example of where this is beneficial?

Reply:
I have used it in test cases where a parameter is being carried and tested under same test, but not in the main code.

Reply:
I don't see the argument for immutable options. Yes, the fluent API allows us to make many of these objects const, but there's nothing in the semantics of `xyz_options` that inherently makes the objects of the type immutable.